### PR TITLE
cgu: added PreCachingConfig resource and unit tests

### PR DESCRIPTION
**Pull request by klaskosk**

### CODE REVIEW

File: pkg/cgu/precachingconfig.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+				Name:      name,
+				Namespace: nsname,
+			},
+		}}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the PreCachingConfig is empty")
+
+		builder.errorMsg = "preCachingConfig 'name' cannot be empty"
+	}
```
Comment (sebrandon1) [inline pkg/cgu/precachingconfig.go:47]:
```suggestion
	if name == "" {
		glog.V(100).Infof("The name of the PreCachingConfig is empty")

		builder.errorMsg = "preCachingConfig 'name' cannot be empty"
		
		return builder
	}
```

File: pkg/cgu/precachingconfig.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the PreCachingConfig is empty")
+
+		builder.errorMsg = "preCachingConfig 'nsname' cannot be empty"
+	}
```
Comment (sebrandon1) [inline pkg/cgu/precachingconfig.go:53]:
```suggestion
	if nsname == "" {
		glog.V(100).Infof("The namespace of the PreCachingConfig is empty")

		builder.errorMsg = "preCachingConfig 'nsname' cannot be empty"
		
		return builder
	}
```

File: pkg/cgu/precachingconfig.go (hunk 1 above)
Comment (sebrandon1) [reply pkg/cgu/precachingconfig.go:47]: Per #355

File: pkg/cgu/precachingconfig.go (hunk 2 above)
Comment (sebrandon1) [reply pkg/cgu/precachingconfig.go:53]: Per #355

File: pkg/cgu/precachingconfig.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the PreCachingConfig is empty")
+
+		return nil, fmt.Errorf("preCachingConfig 'nsname' cannot be empty")
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("preCachingConfig object %s doesn't exist in namespace %s", name, nsname)
```
Comment (sebrandon1) [inline pkg/cgu/precachingconfig.go:91]:
```suggestion
		return nil, fmt.Errorf("preCachingConfig object %s does not exist in namespace %s", name, nsname)
```

Comment (sebrandon1) [reply pkg/cgu/precachingconfig.go:91]: Per #390

File: pkg/cgu/precachingconfig_test.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+			alreadyExists: true,
+			force:         true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testBuilder := buildValidPreCachingConfigTestBuilder(clients.GetTestClients(clients.TestClientParams{}))
+
+		// Create the builder rather than just adding it to the client so that the proper metadata is added and
+		// the update won't fail.
```
Comment (sebrandon1) [inline pkg/cgu/precachingconfig_test.go:258]:
```suggestion
		// the update will not fail.
```

File: pkg/cgu/precachingconfig_test.go [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+			client:                    true,
+			expectedErrorText:         "",
+		},
+		{
+			preCachingConfigName:      defaultPreCachingConfigName,
+			preCachingConfigNamespace: defaultPreCachingConfigNsName,
+			addToRuntimeObjects:       false,
+			client:                    true,
+			expectedErrorText: fmt.Sprintf(
+				"preCachingConfig object %s doesn't exist in namespace %s",
```
Comment (sebrandon1) [inline pkg/cgu/precachingconfig_test.go:73]:
```suggestion
				"preCachingConfig object %s does not exist in namespace %s",
```

**Review by sebrandon1 — COMMENTED** [review]
Some comments

**Comment by klaskosk** [comment]
@sebrandon1 Thanks, I've updated the PR with changes

**Review by kononovn — APPROVED** [review]
/lgtm

**Review by sebrandon1 — APPROVED** [review]

**Comment by sebrandon1** [comment]
Approved!  I don't have the power to resolve my comments above so that will have to be someone else.

**Pull request merged**